src/librustdoc/test.rs

```diff
@@ -703,6 +723,7 @@ impl Tester for Collector {
         let runtool_args = self.options.runtool_args.clone();
         let target = self.options.target.clone();
         let target_str = target.to_string();
+        let visited_tests = Mutex::new(HashMap::new());
```

Comment:
This is creating a new `HashMap` for every test case which is not going to work.

Instead the `HashMap` can be added as a field on `Collector` and updated here in the `add_test` function. That way the generated output directory will be calculated here and passed to `run_test`. This will also avoid the need for a `Mutex`.